File: cpp/linked-list/hard/25.k-个一组翻转链表.cpp
/*
 * @lc app=leetcode.cn id=25 lang=cpp
 *
 * [25] K 个一组翻转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *node = nullptr;
    ListNode *reverseKGroup(ListNode *head, int k)
    {
        auto p = head;
        for (int i = 0; i < k; i++)
        {
            // 不足k个元素,不翻转
            if (p == nullptr)
            {
                return head;
            }
            p = p->next;
        }
        // 翻转head之后的k个元素
        auto newHead = reverseNList(head, k);
        // 翻转之后head变尾结点,链接剩下的链表k组翻转结果
        head->next = reverseKGroup(p, k);
        return newHead;
    }

    ListNode *reverseNList(ListNode *head, int n)
    {
        if (n == 1)
        {
            node = head->next;
            return head;
        }

        auto newHead = reverseNList(head->next, n - 1);
        head->next->next = head;
        head->next = node;
        return newHead;
    }
};
// @lc code=end

File: golang/linked-list/hard/25.K 个一组翻转链表.go
/*
 * @lc app=leetcode.cn id=25 lang=golang
 * @lcpr version=30202
 *
 * [25] K 个一组翻转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
var successor *ListNode

func reverseKGroup(head *ListNode, k int) *ListNode {
	if head == nil {
		return nil
	}

	p := head
	q := head
	for i := 0; i < k; i++ {
		if q == nil {
			return head
		}

		q = q.Next
	}

	newHead := reverseN(p, k)
	head.Next = reverseKGroup(q, k)

	return newHead
}

func reverseN(head *ListNode, n int) *ListNode {
	if n == 1 {
		successor = head.Next
		return head
	}

	newHead := reverseN(head.Next, n-1)
	head.Next.Next = head
	head.Next = successor
	return newHead
}

// @lc code=end

/*
// @lcpr case=start
// [1,2,3,4,5]\n2\n
// @lcpr case=end

// @lcpr case=start
// [1,2,3,4,5]\n3\n
// @lcpr case=end

*/

